### PR TITLE
rewards improvements

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
- math prove 
- single reward claim #759

### CODE REVIEW

File: rewards/src/lib.rs [hunk 1]
```diff
@@ -274,32 +274,77 @@ impl<T: Config> Pallet<T> {
...
+		let total_reward_proportion: T::Balance = U256::from(share.to_owned().saturated_into::<u128>())
+			.saturating_mul(U256::from(total_reward.to_owned().saturated_into::<u128>()))
+			.checked_div(total_shares)
+			.unwrap_or_default()
+			.as_u128()
+			.unique_saturated_into();
+		total_reward_proportion
+			.saturating_sub(withdrawn_reward)
+			.min(total_reward.saturating_sub(*total_withdrawn_reward))
+	}
```
Comment (zqhxuyuan) [inline rewards/src/lib.rs:349]: add `claim_one` method:
```rust
	/// claim one currency for an account
	fn claim_one(
		who: &T::AccountId,
		reward_currency: T::CurrencyId,
		withdrawn_rewards: &mut BTreeMap<T::CurrencyId, T::Balance>,
		share: &mut T::Share,
		total_reward: &mut T::Balance,
		total_shares: U256,
		total_withdrawn_reward: &mut T::Balance,
		pool: &T::PoolId
	) {
		let withdrawn_reward = withdrawn_rewards.get(&reward_currency).copied().unwrap_or_default();

		let total_reward_proportion: T::Balance = U256::from(share.to_owned().saturated_into::<u128>())
			.saturating_mul(U256::from(total_reward.to_owned().saturated_into::<u128>()))
			.checked_div(total_shares)
			.unwrap_or_default()
			.as_u128()
			.unique_saturated_into();
		let reward_to_withdraw = total_reward_proportion
			.saturating_sub(withdrawn_reward)
			.min(total_reward.saturating_sub(*total_withdrawn_reward));
		
		if !reward_to_withdraw.is_zero() {
			*total_withdrawn_reward = total_withdrawn_reward.saturating_add(reward_to_withdraw);
			withdrawn_rewards
				.insert(reward_currency, withdrawn_reward.saturating_add(reward_to_withdraw));

			// pay reward to `who`
			T::Handler::payout(who, pool, reward_currency, reward_to_withdraw);
		}
	}
```

File: rewards/README.md [hunk 2]
```diff
@@ -23,7 +23,7 @@ def inflate(pool, user_share):
 
 def add_share(pool, users, user, user_share):
     # virtually we add more rewards, but claim they were claimed by user
-    # so until `rewards` grows, users will not be able to claim more than zero
+    # so until `rewards` grows, user will not be able to claim more than zero
     to_withdraw = inflate(pool, user_share)
     pool["rewards"] = pool["rewards"] +  to_withdraw
```
Comment (zqhxuyuan) [inline rewards/README.md:28]:
```suggestion
    pool["rewards"] = pool["rewards"] + to_withdraw
```

File: rewards/README.md [hunk 3]
```diff
@@ -23,7 +23,7 @@ def inflate(pool, user_share):
...
     pool["withdrawn_rewards"] = pool["withdrawn_rewards"] +  to_withdraw
```
Comment (zqhxuyuan) [inline rewards/README.md:29]:
```suggestion
    pool["withdrawn_rewards"] = pool["withdrawn_rewards"] + to_withdraw
```

File: rewards/README.md [hunk 4]
```diff
@@ -43,3 +43,31 @@ def claim_rewards(pool, users, user):
...
+
+We want to prove that when a new share is added, it does not dilute previous rewards.
+
+The user who adds a share after the reward is accumulated, will not get any part of the previous reward.
+
+Let $R_n$ be the amount of the current reward asset.
+
+Let $s_i$ be the stake of any specific user our of $m$ total users.
+
+User current reward share equals $$r_i = R_n * ({s_i} / {\sum_{i=1}^m s_i}) $$.
```
Comment (zqhxuyuan) [inline rewards/README.md:57]:
```suggestion
User current reward share equals $$r_i = R_n * ({s_i} / {\sum_{i=1}^m s_i}) $$
```

File: rewards/README.md [hunk 5]
```diff
@@ -43,3 +43,31 @@ def claim_rewards(pool, users, user):
...
+
+User $m + 1$ brings his share, so $$r_i' = R_n * ({s_i} / {\sum_{i=1}^{m+1} s_i}) $$.
```
Comment (zqhxuyuan) [inline rewards/README.md:59]:
```suggestion
User $m + 1$ brings his share, so $$r_i' = R_n * ({s_i} / {\sum_{i=1}^{m+1} s_i}) $$
```

File: rewards/README.md [hunk 6]
```diff
@@ -43,3 +43,31 @@ def claim_rewards(pool, users, user):
...
+
+$r_i > r_i'$, so the original share was diluted and a new user can claim the share of existing users.
+
+What if we increase $R_n$ by $\delta_R$ so that original users get the same share.
+
+We get:
+
+$$ R_n * ({s_i} / {\sum_{i=1}^m s_i}) = ({R_n + \delta_R}) * ({s_i} / {\sum_{i=1}^{m+1} s_i})$$.
```
Comment (zqhxuyuan) [inline rewards/README.md:67]:
```suggestion
$$ R_n * ({s_i} / {\sum_{i=1}^m s_i}) = ({R_n + \delta_R}) * ({s_i} / {\sum_{i=1}^{m+1} s_i})$$
```

**Comment by dzmitry-lahoda** [comment]
cc @zqhxuyuan

**Review by zqhxuyuan — APPROVED** [review]

File: rewards/src/lib.rs [hunk 7]
```diff
@@ -272,34 +272,92 @@ impl<T: Config> Pallet<T> {
...
+			*total_withdrawn_reward = total_withdrawn_reward.saturating_add(reward_to_withdraw);
+			withdrawn_rewards.insert(reward_currency, withdrawn_reward.saturating_add(reward_to_withdraw));
+
+			// pay reward to `who`
+			T::Handler::payout(who, pool, reward_currency, reward_to_withdraw);
+		}
+	}
+
+	fn reward_to_withdraw(
+		share: &mut T::Share,
```
Comment (ermalkaleci) [inline rewards/src/lib.rs:347]: I don't think those parameters should be mutable refs

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs:347]: will check

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs:347]: oh, it was fixed in #770

**Pull request merged**